comment/mixins.py

```diff
@@ -1,162 +1,140 @@
-        self.anonymous = True
-        self.msg = EmailInfo.CONFIRMATION_SENT
 
-    def perform_create(self, comment, request, api=False):
-        if settings.COMMENT_ALLOW_ANONYMOUS and not comment.user:
-            self._handle_anonymous(comment, request, api)
-        else:
-            comment = self.perform_save(comment, request)
-        return comment
+class BaseCommentMixin(BaseCommentPermission):
+    pass
 
 
-class CanCreateMixin(BasePermission, AccessMixin, ValidatorMixin):
+class BaseCreatePermission(UserPermission):
     def has_permission(self, request):
-        return request.user.is_authenticated or settings.COMMENT_ALLOW_ANONYMOUS
+        if not settings.COMMENT_ALLOW_ANONYMOUS and not request.user.is_authenticated:
+            self.reason = ErrorMessage.LOGIN_REQUIRED
+            return False
+        return super().has_permission(request)
 
-    def dispatch(self, request, *args, **kwargs):
-        if not self.has_permission(request):
-            return self.handle_no_permission()
-        return super().dispatch(request, *args, **kwargs)
+
+class CanCreateMixin(BaseCreatePermission, ValidatorMixin):
+    pass
 
 
-class ObjectLevelMixin(BaseCommentMixin):
+class ObjectLevelMixin(BaseCommentPermission):
     @abc.abstractmethod
     def get_object(self):
         raise ImproperlyConfigured(
             ErrorMessage.METHOD_NOT_IMPLEMENTED.format(class_name=self.__class__.__name__, method_name='get_object()')
         )
 
-
-class CanEditMixin(ObjectLevelMixin, ValidatorMixin, abc.ABC):
     def has_object_permission(self, request, obj):
-        return request.user == obj.user
+        return True
 
     def dispatch(self, request, *args, **kwargs):
         obj = self.get_object()
         if not self.has_object_permission(request, obj):
-            return self.handle_no_permission()
+            self.reason = ErrorMessage.NOT_AUTHORIZED
+            data = {'status': self.status, 'reason': self.reason}
+            return UTF8JsonResponse(status=self.status, data=data)
         return super().dispatch(request, *args, **kwargs)
 
 
-class CanDeleteMixin(ObjectLevelMixin, ValidatorMixin, abc.ABC):
+class CanEditMixin(ObjectLevelMixin, ValidatorMixin, abc.ABC):
     def has_object_permission(self, request, obj):
-        return request.user == obj.user or is_comment_admin(request.user) \
-                or (obj.is_flagged and is_comment_moderator(request.user))
+        return request.user == obj.user
 
-    def dispatch(self, request, *args, **kwargs):
-        obj = self.get_object()
-        if not self.has_object_permission(request, obj):
-            return self.handle_no_permission()
-        return super().dispatch(request, *args, **kwargs)
+
+class CanDeleteMixin(ObjectLevelMixin, ValidatorMixin, abc.ABC):
+    def has_object_permission(self, request, obj):
+        return bool(
+            request.user == obj.user or
+            is_comment_admin(request.user) or
```

Comment:
shouldn't these two be part of the `has_permission` method?

Reply:
This is an object level permission. It doesn't make sense moving `is_comment_admin(request.user)` to `has_permission` and keeping `is_comment_moderator(request.user)` here since it's combined with obj check.